substrate/frame/staking/src/lib.rs

```diff
@@ -1223,3 +1223,62 @@ impl BenchmarkingConfig for TestBenchmarkingConfig {
 	type MaxValidators = frame_support::traits::ConstU32<100>;
 	type MaxNominators = frame_support::traits::ConstU32<100>;
 }
+
+/// A result from a disabling decision made by [`DisablingStrategy`].
+pub struct DisablingDecision {
+	/// List of validator indices that should be disabled.
+	disable_offenders: Vec<u32>,
+}
+
+/// Input data for [`make_disabling_decision`]. Provides information about the offence so that the
+/// implementation of [`DisablingStrategy`] can make a decision how to handle the offender.
+pub struct DisablingDecisionContext {
+	pub offender_idx: u32,
+	pub slash_era: EraIndex,
+	pub era_now: EraIndex,
+}
+
+/// Controls validator disabling
+pub trait DisablingStrategy<T: Config> {
+	/// Make a decision if an offender should be disabled or not. The result is an instance of
+	/// `[DisablingDecision]`
+	fn make_disabling_decision(
+		offence_ctx: DisablingDecisionContext,
+		currently_disabled: &Vec<u32>,
+		active_set: &Vec<T::AccountId>,
+	) -> DisablingDecision;
+}
+
+/// Implementation of [`DisablingStrategy`] which disables no more than 1/3 of the validators in the
+/// active set.
+pub struct UpToByzantineThresholdDisablingStrategy;
+
+impl UpToByzantineThresholdDisablingStrategy {
+	/// Disabling limit calculated from the total number of validators in the active set. When
+	/// reached no more validators will be disabled.
+	pub fn byzantine_threshold(validators_len: usize) -> usize {
+		validators_len.saturating_sub(1) / 3
+	}
+}
+
+impl<T: Config> DisablingStrategy<T> for UpToByzantineThresholdDisablingStrategy {
+	fn make_disabling_decision(
+		offence_ctx: DisablingDecisionContext,
+		currently_disabled: &Vec<u32>,
+		active_set: &Vec<T::AccountId>,
+	) -> DisablingDecision {
+		// We don't disable more than 1/3 of the validators in the active set
+		let over_byzantine_threshold =
+			currently_disabled.len() >= Self::byzantine_threshold(active_set.len());
+		// We don't disable for offences in previous eras
+		let ancient_offence = offence_ctx.era_now > offence_ctx.slash_era;
```

Comment:
It doesn't break any current test but I agree it will be a bit harder to unit test. Yet the test can always set the current era in the mock.

If we decide to remove `DisablingDecisionContext` we'd better get the current era from the pallet. If we decide not to - it's fine to pass it externally.

Reply:
>If we decide to remove DisablingDecisionContext we'd better get the current era from the pallet. If we decide not to - it's fine to pass it externally.

Yes, this sounds good to me! I'd go this route to keep things simpler: remove the `DisablingDecisionContext` and access the storage within the trait implementation.

Reply:
Okay, done.